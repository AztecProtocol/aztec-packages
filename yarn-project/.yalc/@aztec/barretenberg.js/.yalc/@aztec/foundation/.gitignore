dest
node_modules
.cache
.env
*.log
*.swp
.tsbuildinfo
.pnp.cjs
.pnp.loader.mjs