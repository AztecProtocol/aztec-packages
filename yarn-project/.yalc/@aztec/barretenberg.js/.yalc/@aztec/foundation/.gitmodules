[submodule "build-system"]
	path = build-system
	url = git@github.com:AztecProtocol/build-system.git
