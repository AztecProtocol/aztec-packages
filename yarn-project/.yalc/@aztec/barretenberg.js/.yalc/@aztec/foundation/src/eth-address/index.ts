import { keccak, randomBytes } from '../crypto/index.js';

export class EthAddress {
  public static SIZE_IN_BYTES = 20;
  public static ZERO = new EthAddress(Buffer.alloc(EthAddress.SIZE_IN_BYTES));

  constructor(private buffer: Buffer) {
    if (buffer.length === 32) {
      if (!buffer.slice(0, 12).equals(Buffer.alloc(12))) {
        throw new Error('Invalid address buffer.');
      } else {
        this.buffer = buffer.slice(12);
      }
    } else if (buffer.length !== EthAddress.SIZE_IN_BYTES) {
      throw new Error(`Expect buffer size to be ${EthAddress.SIZE_IN_BYTES}. Got ${buffer.length}.`);
    }
  }

  public static fromString(address: string) {
    if (!EthAddress.isAddress(address)) {
      throw new Error(`Invalid address string: ${address}`);
    }
    return new EthAddress(Buffer.from(address.replace(/^0x/i, ''), 'hex'));
  }

  public static random() {
    return new EthAddress(randomBytes(20));
  }

  public static isAddress(address: string) {
    if (!/^(0x)?[0-9a-f]{40}$/i.test(address)) {
      // Does not have the basic requirements of an address.
      return false;
    } else if (/^(0x|0X)?[0-9a-f]{40}$/.test(address) || /^(0x|0X)?[0-9A-F]{40}$/.test(address)) {
      // It's ALL lowercase or ALL upppercase.
      return true;
    } else {
      return EthAddress.checkAddressChecksum(address);
    }
  }

  public isZero() {
    return this.equals(EthAddress.ZERO);
  }

  public static checkAddressChecksum(address: string) {
    address = address.replace(/^0x/i, '');
    const addressHash = keccak(Buffer.from(address.toLowerCase())).toString('hex');

    for (let i = 0; i < 40; i++) {
      // The nth letter should be uppercase if the nth digit of casemap is 1.
      if (
        (parseInt(addressHash[i], 16) > 7 && address[i].toUpperCase() !== address[i]) ||
        (parseInt(addressHash[i], 16) <= 7 && address[i].toLowerCase() !== address[i])
      ) {
        return false;
      }
    }
    return true;
  }

  public static toChecksumAddress(address: string) {
    if (!EthAddress.isAddress(address)) {
      throw new Error('Invalid address string.');
    }

    address = address.toLowerCase().replace(/^0x/i, '');
    const addressHash = keccak(Buffer.from(address)).toString();
    let checksumAddress = '0x';

    for (let i = 0; i < address.length; i++) {
      // If ith character is 9 to f then make it uppercase.
      if (parseInt(addressHash[i], 16) > 7) {
        checksumAddress += address[i].toUpperCase();
      } else {
        checksumAddress += address[i];
      }
    }
    return checksumAddress;
  }

  public equals(rhs: EthAddress) {
    return this.buffer.equals(rhs.toBuffer());
  }

  public toString() {
    return '0x' + this.buffer.toString('hex');
  }

  public toChecksumString() {
    return EthAddress.toChecksumAddress(this.buffer.toString('hex'));
  }

  public toBuffer() {
    return this.buffer;
  }

  public toBuffer32() {
    const buffer = Buffer.alloc(32);
    this.buffer.copy(buffer, 12);
    return buffer;
  }
}
