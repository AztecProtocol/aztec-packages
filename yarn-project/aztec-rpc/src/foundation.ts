import nodeCrypto from 'crypto';
import { Keccak } from 'sha3';

export const randomBytes = (len: number) => {
  return nodeCrypto.randomBytes(len) as Buffer;
};

export function keccak256(input: Buffer | string) {
  const inputBuf = typeof input === 'string' ? Buffer.from(input) : input;
  const hash = new Keccak(256);
  return hash.update(inputBuf).digest();
}
