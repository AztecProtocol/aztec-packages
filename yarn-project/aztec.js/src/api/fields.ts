export { Point, Fr, Fq, GrumpkinScalar } from '@aztec/foundation/fields';
