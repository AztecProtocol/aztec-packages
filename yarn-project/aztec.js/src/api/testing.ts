export { AnvilTestWatcher } from '../test/anvil_test_watcher.js';
export { RollupCheatCodes } from '../test/rollup_cheat_codes.js';
export { AztecCheatCodes } from '../test/aztec_cheat_codes.js';
export { EthCheatCodes } from '@aztec/ethereum/eth-cheatcodes';
export { CheatCodes } from '../test/cheat_codes.js';
