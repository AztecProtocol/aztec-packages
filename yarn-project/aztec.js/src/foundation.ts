export function sleep(ms: number) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

export class Timer {
  private start: number;

  constructor() {
    this.start = new Date().getTime();
  }

  public ms() {
    return new Date().getTime() - this.start;
  }

  public s() {
    return (new Date().getTime() - this.start) / 1000;
  }
}

export function* backoffGenerator() {
  const v = [1, 1, 1, 2, 4, 8, 16, 32, 64];
  let i = 0;
  while (true) {
    yield v[Math.min(i++, v.length - 1)];
  }
}

export async function retry<Result>(fn: () => Promise<Result>, name = 'Operation', backoff = backoffGenerator()) {
  while (true) {
    try {
      return await fn();
    } catch (err: any) {
      const s = backoff.next().value;
      if (s === undefined) {
        throw err;
      }
      console.log(`${name} failed. Will retry in ${s}s...`);
      console.log(err);
      await sleep(s * 1000);
      continue;
    }
  }
}

// Call `fn` repeatedly until it returns true or timeout.
// Both `interval` and `timeout` are seconds.
// Will never timeout if the value is 0.
export async function retryUntil<T>(fn: () => Promise<T | undefined>, name = '', timeout = 0, interval = 1) {
  const timer = new Timer();
  while (true) {
    const result = await fn();
    if (result) {
      return result;
    }

    await sleep(interval * 1000);

    if (timeout && timer.s() > timeout) {
      throw new Error(name ? `Timeout awaiting ${name}` : 'Timeout');
    }
  }
}
