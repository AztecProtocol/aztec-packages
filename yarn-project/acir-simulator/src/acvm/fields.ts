import { AztecAddress, EthAddress, Fr } from '@aztec/foundation';

export type ACVMField = `0x${string}`;
export type ACVMWitness = Map<number, ACVMField>;

export const ZERO_ACVM_FIELD: ACVMField = `0x${Buffer.alloc(32).toString('hex')}`;
export const ONE_ACVM_FIELD: ACVMField = `0x${'00'.repeat(31)}01`;

function adaptBufferSize(originalBuf: Buffer) {
  const buffer = Buffer.alloc(32);
  if (originalBuf.length > buffer.length) {
    throw new Error('Buffer does not fit in 32 bytes');
  }
  originalBuf.copy(buffer, buffer.length - originalBuf.length);
  return buffer;
}

export function toACVMField(value: AztecAddress | EthAddress | Fr | Buffer | boolean | number): `0x${string}` {
  if (typeof value === 'boolean') {
    return value ? ONE_ACVM_FIELD : ZERO_ACVM_FIELD;
  }

  let buffer;

  if (Buffer.isBuffer(value)) {
    buffer = value;
  } else if (typeof value === 'number') {
    buffer = Buffer.alloc(32);
    buffer.writeUInt32BE(value, 28);
  } else {
    buffer = value.toBuffer();
  }

  return `0x${adaptBufferSize(buffer).toString('hex')}`;
}

export function fromACVMField(field: `0x${string}`): Fr {
  const buffer = Buffer.from(field.slice(2), 'hex');
  return Fr.fromBuffer(buffer);
}
