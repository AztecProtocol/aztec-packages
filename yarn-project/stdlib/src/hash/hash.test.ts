import { times } from '@aztec/foundation/collection';
import { EthAddress } from '@aztec/foundation/eth-address';
import { Fr } from '@aztec/foundation/fields';
import { setupCustomSnapshotSerializers } from '@aztec/foundation/testing';
import { updateInlineTestData } from '@aztec/foundation/testing/files';

import { AztecAddress } from '../aztec-address/index.js';
import { makeAztecAddress } from '../tests/factories.js';
import {
  computeL2ToL1MessageHash,
  computeNoteHashNonce,
  computePublicDataTreeLeafSlot,
  computePublicDataTreeValue,
  computeSecretHash,
  computeUniqueNoteHash,
  computeVarArgsHash,
  siloNoteHash,
  siloNullifier,
} from './hash.js';

describe('hash', () => {
  setupCustomSnapshotSerializers(expect);

  it('computes note hash nonce', async () => {
    const nullifierZero = new Fr(123n);
    const noteHashIndex = 456;
    const res = await computeNoteHashNonce(nullifierZero, noteHashIndex);
    expect(res).toMatchSnapshot();
  });

  it('computes unique note hash', async () => {
    const noteNonce = new Fr(123n);
    const noteHash = new Fr(456);
    const res = await computeUniqueNoteHash(noteNonce, noteHash);
    expect(res).toMatchSnapshot();
  });

  it('computes siloed note hash', async () => {
    const contractAddress = new AztecAddress(new Fr(123n).toBuffer());
    const uniqueNoteHash = new Fr(456);
    const res = await siloNoteHash(contractAddress, uniqueNoteHash);
    expect(res).toMatchSnapshot();
  });

  it('computes siloed nullifier', async () => {
    const contractAddress = new AztecAddress(new Fr(123n).toBuffer());
    const innerNullifier = new Fr(456);
    const res = await siloNullifier(contractAddress, innerNullifier);
    expect(res).toMatchSnapshot();
  });

  it('computes public data tree value', () => {
    const value = new Fr(3n);
    const res = computePublicDataTreeValue(value);
    expect(res).toMatchSnapshot();
  });

  it('computes public data tree leaf slot', async () => {
    const contractAddress = makeAztecAddress();
    const value = new Fr(3n);
    const res = await computePublicDataTreeLeafSlot(contractAddress, value);
    expect(res).toMatchSnapshot();
  });

  it('hashes empty function args', async () => {
    const res = await computeVarArgsHash([]);
    expect(res).toMatchSnapshot();
  });

  it('hashes function args', async () => {
    const args = times(8, i => new Fr(i));
    const res = await computeVarArgsHash(args);
    expect(res).toMatchSnapshot();
  });

  it('hashes many function args', async () => {
    const args = times(200, i => new Fr(i));
    const res = await computeVarArgsHash(args);
    expect(res).toMatchSnapshot();
  });

  it('compute secret message hash', async () => {
    const value = new Fr(8n);
    const hash = await computeSecretHash(value);
    expect(hash).toMatchSnapshot();
  });

  it('Var args hash matches noir', async () => {
    const args = times(100, i => new Fr(i));
    const res = await computeVarArgsHash(args);
    expect(res).toMatchSnapshot();

    // Value used in "compute_var_args_hash" test in hash.nr
    // console.log("hash", hash);
  });

  it('L2ToL1Message siloing matches Noir', () => {
    const nonEmptyHash = computeL2ToL1MessageHash({
      l2Sender: AztecAddress.fromField(new Fr(3)),
      l1Recipient: EthAddress.fromField(new Fr(1)),
      content: new Fr(2),
      rollupVersion: new Fr(4),
      chainId: new Fr(5),
    });

    expect(nonEmptyHash.toString()).toMatchInlineSnapshot(
      `"0x0081edf209e087ad31b3fd24263698723d57190bd1d6e9fe056fc0c0a68ee661"`,
    );

    // Run with AZTEC_GENERATE_TEST_DATA=1 to update noir test data
    updateInlineTestData(
      'noir-projects/noir-protocol-circuits/crates/types/src/hash.nr',
      'hash_from_typescript',
      nonEmptyHash.toString(),
    );
  });
});
