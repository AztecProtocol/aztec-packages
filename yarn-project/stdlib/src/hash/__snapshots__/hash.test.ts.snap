// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hash Var args hash matches noir 1`] = `Fr<0x19b0d74feb06ebde19edd85a28986c97063e84b3b351a8b666c7cac963ce655f>`;

exports[`hash compute secret message hash 1`] = `Fr<0x211b950e538ee911389e441942d6646d13d725957a49d06500ae9bb673b0fbc2>`;

exports[`hash computes note hash nonce 1`] = `Fr<0x24cbf5ece05503c37381b5d7dfcaf96fe2aca3749cb1a5d4d2f5264e40872fa2>`;

exports[`hash computes public data tree leaf slot 1`] = `Fr<0x13c520847ef2069f11086e5085b756d5e03931879b7e6a63eca639e4b5ad4a4a>`;

exports[`hash computes public data tree value 1`] = `Fr<0x0000000000000000000000000000000000000000000000000000000000000003>`;

exports[`hash computes siloed note hash 1`] = `Fr<0x12ddf0068d9dba68cb857f9b431064c0c04c06fad060b9df25894ccc4a790066>`;

exports[`hash computes siloed nullifier 1`] = `Fr<0x280270c4a9431f6d4f029f5cd77807c5d797dab9e83181a47a9082ae2c5a3149>`;

exports[`hash computes unique note hash 1`] = `Fr<0x2f9b80d22188c5383a577596538f71f9713dc220bf36b6497ec16ea0f9df2282>`;

exports[`hash hashes empty function args 1`] = `Fr<0x0000000000000000000000000000000000000000000000000000000000000000>`;

exports[`hash hashes function args 1`] = `Fr<0x08fc05646caf2d63a8f3b424351783a521edaf28da0933d7c5b78a27e7497dac>`;

exports[`hash hashes many function args 1`] = `Fr<0x03933d5a59efde5af91c47a93e6c4965f0506d160f76c701b2de1503d74c7775>`;
